pkg/pma/pma.c

```diff
@@ -0,0 +1,817 @@
+{
+    size_t  pg_idx    = _addr_to_page_idx(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t status    = (pma->pg_status[entry_idx] >> bit_idx) & PS_MASK;
+    return status;
+}
+
+/// Set the status of the page surrounding an address. To get rather than set,
+/// see _page_status().
+///
+/// @param[in] addr    Address within the page in question. Must be within the
+///                    bounds of the PMA.
+/// @param[in] status  New status of the page in question.
+/// @param[in] pma     PMA the page in question belongs to.
+static inline void
+_set_page_status(const void *addr, page_status_t status, const pma_t *pma)
+{
+    size_t  pg_idx    = _addr_to_page_idx(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t entry     = pma->pg_status[entry_idx];
+    pma->pg_status[entry_idx]
+        = (entry & ~(PS_MASK << bit_idx)) | (status << bit_idx);
+}
+
+/// Set the status of the page range starting at an address. To set the page
+/// status of a single page, see _set_page_status().
+///
+/// @param[in] addr    Address within the first page in question. Must be within
+///                    the bounds of the PMA.
+/// @param[in] pg_cnt  Number of pages in the range. The entirety of the page
+///                    range must be within the bounds of the PMA.
+/// @param[in] status  New status of the page range in question.
+/// @param[in] pma     PMA the page range in question belongs to.
+static inline void
+_set_page_status_range(const void   *addr,
+                       size_t        pg_cnt,
+                       page_status_t status,
+                       const pma_t  *pma)
+{
+    for (size_t i = 0; i < pg_cnt * kPageSz; i += kPageSz) {
+        _set_page_status((char *)addr + i, status, pma);
+    }
+}
+
+/// Sync changes made to either the heap or stack since the last call to
+/// pma_sync().
+/// Append dirty pages (pages modified since the last call to pma_sync()) to a
+/// write-ahead log.
+///
+/// @param[in] wal         Write-ahead log. Must not be NULL.
+/// @param[in] base        Base address of the file-backed mapping.
+/// @param[in] grows_down  true if the mapping grows downward in memory.
+/// @param[in] pma         PMA this mapping belongs to.
+/// @param[in] len         Number of bytes at the beginning of the mapping to be
+///                        synced to the backing file. Must be a multiple of
+///                        kPageSz.
+/// @param[in] fd          Open file descriptor of the backing file.
+///
+/// @return 0   Synced changes successfully.
+/// @return -1  Failed to sync changes.
+static int
+_append_dirty_pages(wal_t *wal,
+                    void  *base,
+                    bool   grows_down,
+                    pma_t *pma,
+                    size_t len,
+                    int    fd);
```

Comment:
`fd` param is unused.

Reply:
That should have produced a compiler warning unless we accidentally disabled that somehow. I think warn on unused local var is disabled.

Reply:
Fixed. 

I looked into turning on `-Wunused-parameter`, but doing so produces a ton of warnings when building `h2o`. I think we should straighten out the warning flags in a different PR.

Reply:
Good idea. Should also probably fixup the project to get it into a state where we can warn on unused local variables as well. Last I checked, there weren't that many instances to fix.